File: web-app/src/services/aiService.js
import api from './api';

/**
 * AI Service for IBCM AI-powered features and recommendations
 * Integrates with IBCM-ai microservice through the backend API
 */
class AIService {
  /**
   * Perform enhanced search with AI capabilities
   * @param {string} query - Search query
   * @param {string} userId - User ID
   * @param {Object} filters - Optional filters
   * @param {Object} location - Optional location object
   * @param {Object} preferences - Optional user preferences
   * @returns {Promise<Object>} Search results with metadata
   */
  async enhancedSearch(query, userId, filters = null, location = null, preferences = null) {
    try {
      const response = await api.post('/api/ai/search', {
        query,
        userId,
        filters,
        location,
        preferences
      });
      return response.data;
    } catch (error) {
      console.error('Enhanced search error:', error);
      throw error;
    }
  }

  /**
   * Get personalized recommendations for a user
   * @param {string} userId - User ID
   * @param {number} limit - Number of recommendations to return
   * @param {string} type - Optional type filter
   * @param {string} location - Optional location
   * @returns {Promise<Array>} List of recommendations
   */
  async getPersonalizedRecommendations(userId, limit = 10, type = null, location = null) {
    try {
      let url = `/api/ai/recommendations?userId=${userId}&limit=${limit}`;
      if (type) url += `&type=${type}`;
      if (location) url += `&location=${encodeURIComponent(location)}`;
      
      const response = await api.get(url);
      return response.data;
    } catch (error) {
      console.error('Personalized recommendations error:', error);
      // Return empty array as fallback
      if (error.response?.status === 404 || !error.response) {
        return [];
      }
      throw error;
    }
  }

  /**
   * Get optimized recommendations based on context
   * @param {string} userId - User ID
   * @param {string} context - Optional context
   * @param {number} limit - Number of recommendations to return
   * @returns {Promise<Array>} List of recommendations
   */
  async getOptimizedRecommendations(userId, context = null, limit = 10) {
    try {
      let url = `/api/ai/recommendations/optimized?userId=${userId}&limit=${limit}`;
      if (context) url += `&context=${encodeURIComponent(context)}`;
      
      const response = await api.get(url);
      return response.data;
    } catch (error) {
      console.error('Optimized recommendations error:', error);
      // Return empty array as fallback
      if (error.response?.status === 404 || !error.response) {
        return [];
      }
      throw error;
    }
  }

  /**
   * Get location-based recommendations
   * @param {string} userId - User ID
   * @param {number} latitude - Latitude
   * @param {number} longitude - Longitude
   * @param {number} radius - Search radius in km
   * @param {string} category - Optional category
   * @param {number} limit - Number of recommendations to return
   * @returns {Promise<Array>} List of recommendations
   */
  async getLocationRecommendations(userId, latitude, longitude, radius = 10, category = null, limit = 10) {
    try {
      const response = await api.post('/api/ai/location-recommendations', {
        userId,
        latitude,
        longitude,
        radius,
        category,
        limit
      });
      return response.data;
    } catch (error) {
      console.error('Location recommendations error:', error);
      // Return empty array as fallback
      if (error.response?.status === 404 || !error.response) {
        return [];
      }
      throw error;
    }
  }

  /**
   * Get user insights and analytics
   * @param {string} userId - User ID
   * @param {string} period - Time period (day, week, month, year)
   * @returns {Promise<Object>} User insights
   */
  async getUserInsights(userId, period = 'month') {
    try {
      const response = await api.get(`/api/ai/user/insights?userId=${userId}&period=${period}`);
      return response.data;
    } catch (error) {
      console.error('User insights error:', error);
      throw error;
    }
  }

  /**
   * Get user preferences
   * @param {string} userId - User ID
   * @returns {Promise<Object>} User preferences
   */
  async getUserPreferences(userId) {
    try {
      const response = await api.get(`/api/ai/user/preferences?userId=${userId}`);
      return response.data;
    } catch (error) {
      console.error('User preferences error:', error);
      throw error;
    }
  }

  /**
   * Update user preferences
   * @param {string} userId - User ID
   * @param {Object} preferences - User preferences
   * @returns {Promise<Object>} Updated user preferences
   */
  async updateUserPreferences(userId, preferences) {
    try {
      const response = await api.put('/api/ai/user/preferences', {
        userId,
        preferences
      });
      return response.data;
    } catch (error) {
      console.error('Update user preferences error:', error);
      throw error;
    }
  }

  /**
   * Generate event description
   * @param {string} title - Event title
   * @param {string} category - Event category
   * @param {string} context - Optional context
   * @returns {Promise<Object>} Generated content
   */
  async generateEventDescription(title, category, context = '') {
    try {
      const response = await api.post('/api/ai/generate-description', {
        title,
        category,
        context
      });
      return response.data;
    } catch (error) {
      console.error('Generate event description error:', error);
      throw error;
    }
  }

  /**
   * Generate tags for content
   * @param {string} content - Content to generate tags for
   * @param {number} count - Number of tags to generate
   * @returns {Promise<Array>} List of tags
   */
  async generateTags(content, count = 5) {
    try {
      const response = await api.post('/api/ai/generate-tags', {
        content,
        count
      });
      return response.data;
    } catch (error) {
      console.error('Generate tags error:', error);
      // Return empty array as fallback
      if (error.response?.status === 404 || !error.response) {
        return [];
      }
      throw error;
    }
  }

  /**
   * Analyze sentiment of text
   * @param {string} text - Text to analyze
   * @returns {Promise<Object>} Sentiment analysis
   */
  async analyzeSentiment(text) {
    try {
      const response = await api.post('/api/ai/sentiment-analysis', {
        text
      });
      return response.data;
    } catch (error) {
      console.error('Sentiment analysis error:', error);
      throw error;
    }
  }

  /**
   * Submit feedback for AI learning
   * @param {string} userId - User ID
   * @param {string} targetId - Target ID
   * @param {number} feedback - Feedback score
   * @param {string} comment - Optional comment
   * @param {string} type - Feedback type
   * @returns {Promise<string>} Success message
   */
  async submitFeedback(userId, targetId, feedback, comment = null, type) {
    try {
      const response = await api.post('/api/ai/feedback', {
        userId,
        targetId,
        feedback,
        comment,
        type
      });
      return response.data;
    } catch (error) {
      console.error('Submit feedback error:', error);
      throw error;
    }
  }

  /**
   * Get business analytics
   * @param {string} businessId - Business ID
   * @param {string} period - Time period
   * @returns {Promise<Object>} Business analytics
   */
  async getBusinessAnalytics(businessId, period = 'month') {
    try {
      const response = await api.get(`/api/ai/business-analytics?businessId=${businessId}&period=${period}`);
      return response.data;
    } catch (error) {
      console.error('Business analytics error:', error);
      throw error;
    }
  }

  /**
   * Generate marketing campaign
   * @param {string} businessId - Business ID
   * @param {string} eventId - Event ID
   * @param {string} context - Context
   * @returns {Promise<Object>} Generated campaign
   */
  async generateCampaign(businessId, eventId, context) {
    try {
      const response = await api.post('/api/ai/generate-campaign', {
        businessId,
        eventId,
        context
      });
      return response.data;
    } catch (error) {
      console.error('Generate campaign error:', error);
      throw error;
    }
  }
}

export default new AIService(); 

File: IBCM-ios/IBCM/IBCM/Services/AIService.swift
import Foundation
import Combine

/**
 * AI Service for IBCM AI-powered features and recommendations
 * Integrates with IBCM-ai microservice through the backend API
 */
class AIService: BaseAPIService {
    static let shared = AIService()
    
    private init() {}
    
    // MARK: - Search and Recommendations
    
    func enhancedSearch(query: String, userId: String, filters: [String: Any]? = nil, 
                        location: Location? = nil, preferences: [String: Any]? = nil) async throws -> SearchResponse {
        let request = EnhancedSearchRequest(
            query: query,
            userId: userId,
            filters: filters,
            location: location,
            preferences: preferences
        )
        
        return try await apiService.request(
            endpoint: "api/ai/search",
            method: HTTPMethod.post.rawValue,
            body: try? JSONEncoder().encode(request)
        )
    }
    
    func getPersonalizedRecommendations(userId: String, limit: Int = 10, 
                                        type: String? = nil, location: String? = nil) async throws -> [RecommendationItem] {
        var queryItems = [
            URLQueryItem(name: "userId", value: userId),
            URLQueryItem(name: "limit", value: "\(limit)")
        ]
        
        if let type = type {
            queryItems.append(URLQueryItem(name: "type", value: type))
        }
        
        if let location = location {
            queryItems.append(URLQueryItem(name: "location", value: location))
        }
        
        return try await apiService.request(
            endpoint: "api/ai/recommendations",
            method: HTTPMethod.get.rawValue,
            queryItems: queryItems
        )
    }
    
    func getOptimizedRecommendations(userId: String, context: String? = nil, 
                                     limit: Int = 10) async throws -> [RecommendationItem] {
        var queryItems = [
            URLQueryItem(name: "userId", value: userId),
            URLQueryItem(name: "limit", value: "\(limit)")
        ]
        
        if let context = context {
            queryItems.append(URLQueryItem(name: "context", value: context))
        }
        
        return try await apiService.request(
            endpoint: "api/ai/recommendations/optimized",
            method: HTTPMethod.get.rawValue,
            queryItems: queryItems
        )
    }
    
    func getLocationRecommendations(userId: String, latitude: Double, longitude: Double,
                                   radius: Double = 10.0, category: String? = nil, 
                                   limit: Int = 10) async throws -> [RecommendationItem] {
        let request = LocationRecommendationRequest(
            userId: userId,
            latitude: latitude,
            longitude: longitude,
            radius: radius,
            category: category,
            limit: limit
        )
        
        return try await apiService.request(
            endpoint: "api/ai/location-recommendations",
            method: HTTPMethod.post.rawValue,
            body: try? JSONEncoder().encode(request)
        )
    }
    
    // MARK: - User Insights and Analytics
    
    func getUserInsights(userId: String, period: String = "month") async throws -> UserInsights {
        let queryItems = [
            URLQueryItem(name: "userId", value: userId),
            URLQueryItem(name: "period", value: period)
        ]
        
        return try await apiService.request(
            endpoint: "api/ai/user/insights",
            method: HTTPMethod.get.rawValue,
            queryItems: queryItems
        )
    }
    
    func getUserPreferences(userId: String) async throws -> UserPreferences {
        let queryItems = [
            URLQueryItem(name: "userId", value: userId)
        ]
        
        return try await apiService.request(
            endpoint: "api/ai/user/preferences",
            method: HTTPMethod.get.rawValue,
            queryItems: queryItems
        )
    }
    
    func updateUserPreferences(userId: String, preferences: [String: Any]) async throws -> UserPreferences {
        let request = UpdatePreferencesRequest(userId: userId, preferences: preferences)
        
        return try await apiService.request(
            endpoint: "api/ai/user/preferences",
            method: HTTPMethod.put.rawValue,
            body: try? JSONEncoder().encode(request)
        )
    }
    
    // MARK: - Content Generation
    
    func generateEventDescription(title: String, category: String, context: String = "") async throws -> GeneratedContent {
        let request = GenerateDescriptionRequest(title: title, category: category, context: context)
        
        return try await apiService.request(
            endpoint: "api/ai/generate-description",
            method: HTTPMethod.post.rawValue,
            body: try? JSONEncoder().encode(request)
        )
    }
    
    func generateTags(content: String, count: Int = 5) async throws -> [String] {
        let request = GenerateTagsRequest(content: content, count: count)
        
        return try await apiService.request(
            endpoint: "api/ai/generate-tags",
            method: HTTPMethod.post.rawValue,
            body: try? JSONEncoder().encode(request)
        )
    }
    
    // MARK: - Sentiment Analysis
    
    func analyzeSentiment(text: String) async throws -> SentimentAnalysis {
        let request = SentimentAnalysisRequest(text: text)
        
        return try await apiService.request(
            endpoint: "api/ai/sentiment-analysis",
            method: HTTPMethod.post.rawValue,
            body: try? JSONEncoder().encode(request)
        )
    }
    
    // MARK: - Feedback and Learning
    
    func submitFeedback(userId: String, targetId: String, feedback: Int, 
                       comment: String? = nil, type: String) async throws -> String {
        let request = AIFeedbackRequest(
            userId: userId,
            targetId: targetId,
            feedback: feedback,
            comment: comment,
            type: type
        )
        
        return try await apiService.request(
            endpoint: "api/ai/feedback",
            method: HTTPMethod.post.rawValue,
            body: try? JSONEncoder().encode(request)
        )
    }
}

// MARK: - Request Models

struct EnhancedSearchRequest: Codable {
    let query: String
    let userId: String
    let filters: [String: Any]?
    let location: Location?
    let preferences: [String: Any]?
    let context: String?
    
    init(query: String, userId: String, filters: [String: Any]? = nil, 
         location: Location? = nil, preferences: [String: Any]? = nil, context: String? = nil) {
        self.query = query
        self.userId = userId
        self.filters = filters
        self.location = location
        self.preferences = preferences
        self.context = nil
    }
    
    enum CodingKeys: String, CodingKey {
        case query, userId, location, context
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(query, forKey: .query)
        try container.encode(userId, forKey: .userId)
        try container.encodeIfPresent(location, forKey: .location)
        try container.encodeIfPresent(context, forKey: .context)
        
        // Handle dictionary encoding manually
        if let filters = filters {
            let data = try JSONSerialization.data(withJSONObject: filters)
            let filterDict = try JSONSerialization.jsonObject(with: data)
            try container.encode(filterDict, forKey: .filters)
        }
        
        if let preferences = preferences {
            let data = try JSONSerialization.data(withJSONObject: preferences)
            let preferencesDict = try JSONSerialization.jsonObject(with: data)
            try container.encode(preferencesDict, forKey: .preferences)
        }
    }
}

struct LocationRecommendationRequest: Codable {
    let userId: String
    let latitude: Double
    let longitude: Double
    let radius: Double
    let category: String?
    let limit: Int
}

struct GenerateDescriptionRequest: Codable {
    let title: String
    let category: String
    let context: String
}

struct GenerateTagsRequest: Codable {
    let content: String
    let count: Int
}

struct SentimentAnalysisRequest: Codable {
    let text: String
}

struct AIFeedbackRequest: Codable {
    let userId: String
    let targetId: String
    let feedback: Int
    let comment: String?
    let type: String
}

struct UpdatePreferencesRequest: Codable {
    let userId: String
    let preferences: [String: Any]
    
    enum CodingKeys: String, CodingKey {
        case userId, preferences
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(userId, forKey: .userId)
        
        // Handle dictionary encoding manually
        let data = try JSONSerialization.data(withJSONObject: preferences)
        let preferencesDict = try JSONSerialization.jsonObject(with: data)
        try container.encode(preferencesDict, forKey: .preferences)
    }
}

// MARK: - Response Models

struct SearchResponse: Codable {
    let results: [SearchResult]
    let metadata: SearchMetadata?
}

struct SearchResult: Codable, Identifiable {
    let id: String
    let title: String
    let description: String
    let type: String
    let score: Double
    let metadata: [String: String]?
}

struct SearchMetadata: Codable {
    let totalResults: Int
    let processingTime: Double
    let queryContext: String?
}

struct RecommendationItem: Codable, Identifiable {
    let id: String
    let title: String
    let description: String
    let type: String
    let score: Double
    let reasons: [String]?
    let imageUrl: String?
    let metadata: [String: String]?
}

struct UserInsights: Codable {
    let preferences: [String: Double]
    let interests: [String]
    let behavior: BehaviorMetrics
    let recommendations: [RecommendationInsight]
}

struct BehaviorMetrics: Codable {
    let engagementScore: Double
    let activityLevel: String
    let lastActive: Date
    let frequentTimes: [String]
}

struct RecommendationInsight: Codable {
    let category: String
    let confidence: Double
    let reason: String
}

struct UserPreferences: Codable {
    let categories: [String]
    let locations: [String]
    let priceRange: PriceRange
    let notificationSettings: NotificationPreferences
    let customPreferences: [String: String]
}

struct PriceRange: Codable {
    let min: Double
    let max: Double
}

struct NotificationPreferences: Codable {
    let enabled: Bool
    let types: [String]
    let frequency: String
}

struct GeneratedContent: Codable {
    let content: String
    let metadata: ContentMetadata
}

struct ContentMetadata: Codable {
    let wordCount: Int
    let readingTime: Int
    let keywords: [String]
    let sentiment: String
}

struct SentimentAnalysis: Codable {
    let sentiment: String
    let score: Double
    let confidence: Double
    let aspects: [AspectSentiment]?
}

struct AspectSentiment: Codable {
    let aspect: String
    let sentiment: String
    let score: Double
} 